crates/block-producer/src/transaction.rs

```diff
@@ -0,0 +1,131 @@
+    }
+
+    pub fn account_update(&self) -> &TxAccountUpdate {
+        self.inner.0.account_update()
+    }
+
+    pub fn store_account_state(&self) -> Option<Digest> {
+        self.store_account_state
+    }
+
+    pub fn authentication_height(&self) -> u32 {
+        self.authentication_height
+    }
+
+    pub fn nullifiers(&self) -> impl Iterator<Item = Nullifier> + '_ {
+        self.inner.0.get_nullifiers()
+    }
+
+    pub fn output_notes(&self) -> impl Iterator<Item = NoteId> + '_ {
+        self.inner.0.output_notes().iter().map(|note| note.id())
+    }
+
+    pub fn unauthenticated_notes(&self) -> impl Iterator<Item = NoteId> + '_ {
+        self.inner
+            .0
+            .get_unauthenticated_notes()
+            .cloned()
+            .map(|header| header.id())
+            .filter(|note_id| !self.authenticated_notes.contains(note_id))
+    }
+
+    pub fn into_raw(self) -> ProvenTransaction {
+        self.inner.0
+    }
+}
+
+#[cfg(test)]
+impl AuthenticatedTransaction {
+    pub fn from_inner(inner: ProvenTransaction) -> Self {
+        let store_account_state = match inner.account_update().init_state_hash() {
+            zero if zero == Digest::default() => None,
+            non_zero => Some(non_zero),
+        };
+        Self {
+            inner: VerifiedTransaction::new_unchecked(inner),
+            store_account_state,
+            authenticated_notes: Default::default(),
+            authentication_height: Default::default(),
+        }
+    }
+
+    pub fn with_store_state(mut self, state: Digest) -> Self {
+        self.store_account_state = Some(state);
+        self
+    }
+
+    pub fn with_empty_store_state(mut self) -> Self {
+        self.store_account_state = None;
+        self
+    }
+}
+
+impl VerifiedTransaction {
+    /// Creates a new verified transaction without actually verifying the proof.
+    ///
+    /// The caller assumes reponsibility for ensuring the proof was actually valid.
+    pub fn new_unchecked(tx: ProvenTransaction) -> Self {
+        Self(tx)
+    }
+
+    /// Validates the transaction against the inputs from the store.
+    pub fn validate_inputs(
+        self,
+        inputs: TransactionInputs,
+    ) -> Result<AuthenticatedTransaction, VerifyTxError> {
+        let nullifiers_already_spent = self
+            .0
+            .get_nullifiers()
+            .filter(|nullifier| inputs.nullifiers.get(nullifier).cloned().flatten().is_some())
+            .collect::<Vec<_>>();
+        if !nullifiers_already_spent.is_empty() {
+            return Err(VerifyTxError::InputNotesAlreadyConsumed(nullifiers_already_spent));
+        }
+
+        // Invert the missing notes; i.e. we now know the rest were actually found.
+        let authenticated_notes = self
+            .0
+            .get_unauthenticated_notes()
+            .map(|header| header.id())
+            .filter(|note_id| !inputs.missing_unauthenticated_notes.contains(note_id))
+            .collect();
```

Comment:
We should probably change how `GetTransactionInputs` endpoint works so that it returns newly authenticated notes rather than missing notes. Let's create an issue for this.

Reply:
I was thinking we should actually return the inclusion proofs, and carry those with.

With the goal being less store IO at the time of batch/block proving as we already have some of the things. Though maybe this isn't worth it.

Reply:
#519

Reply:
> I was thinking we should actually return the inclusion proofs, and carry those with.
> 
> With the goal being less store IO at the time of batch/block proving as we already have some of the things. Though maybe this isn't worth it.

My thinking here was that since we'd need to go to the store anyways at the time of batch/block building, it's probably not worth it to get the paths immediately - especially since we don't know if the transaction is valid yet (i.e., it hasn't been checked against the in-flight state).